### ISSUE TITLE
Provide help for common code generation issues

### ISSUE THREAD

**Issue by mdgood** [w3c/musicxml#280]
The core features of MusicXML were designed before W3C XSD Schemas were in wide use. This means they were designed without being able to take into account the needs of code generation tools like JAXB that are based on XSD definitions.

Several issues have come up involving difficulties in using code generation tools due to these limitations (#209, #210, #256, #257). When adding new features to MusicXML we can design around these limitations as we did in #209. But many issues cannot be addressed due to the need to maintain compatibility with previous MusicXML versions.

Many of these issues have some common solutions, such as using a JAXB bindings file. We could provide a standard JAXB bindings file as part of the MusicXML specification, and similar files for other common code generation tools. We could also document these issues and suggested solutions for tools where we do not provide a standard solution.

**Comment by jsutdolph** [comment]
I found a solution that works for me to the grace cue problem #210.

Modified musicxml.xsd to restore note grace to 3.0 definition:

ie
	<xs:sequence>
		<xs:element name="grace" type="grace"/>
		<xs:choice>
			<xs:sequence>
				<xs:group ref="full-note"/>
				<xs:element name="tie" type="tie" minOccurs="0" maxOccurs="2"/>
			</xs:sequence>
			<xs:sequence>
				<xs:element name="cue" type="empty"/>
				<xs:group ref="full-note"/>
			</xs:sequence>
		</xs:choice>
	</xs:sequence>

is restored to:

	<xs:sequence>
		<xs:element name="grace" type="grace"/>
		<xs:group ref="full-note"/>
		<xs:element name="tie" type="tie" minOccurs="0" maxOccurs="2"/>
	</xs:sequence>


2.  Add optional cue element to grace so we support grace cue.
The only thing we lose is that a tie on a grace cue is not a validation error

Now:
	<xs:sequence>
		<xs:element name="grace" type="grace"/>
		<xs:element name="cue" type="empty" minOccurs="0" maxOccurs="1"/>
		<xs:group ref="full-note"/>
		<xs:element name="tie" type="tie" minOccurs="0" maxOccurs="2"/>
	</xs:sequence>

**Comment by jsutdolph** [comment]
SeeScore uses xsd/e from CodeSynthesis https://www.codesynthesis.com/products/xsde/
Note that it is important for the parser to preserve the ordering of elements.

**Comment by webern** [comment]
This is a great idea. I ended up pseudo-generating my own C++ code because a lot of tooling can’t seem to cope with MusicXML. Consolidating knowledge about what works, what doesn’t, and why would be really helpful.

**Comment by dancarpenter21** [comment]
The following "bindings.xjb" file allowed me to generate Java classes with JDK 8 and `xjc ./musicxml.xsd -d ./src -b ./bindings.xjb` from a terminal.

`<bindings xmlns="http://java.sun.com/xml/ns/jaxb"
          xmlns:xsi="http://www.w3.org/2000/10/XMLSchema-instance"
          xmlns:xs="http://www.w3.org/2001/XMLSchema"
          version="2.1">
    <bindings schemaLocation="musicxml.xsd" version="1.0">
        
        <schemaBindings>
            <package name="org.musicxml_3_1"/>
        </schemaBindings>

        <bindings node="//xs:complexType[@name='barline']">
            <bindings node=".//xs:attribute[@name='segno']">
                <property name="segnoAttribute"/>
            </bindings>
        </bindings>

        <bindings node="//xs:complexType[@name='barline']">
            <bindings node=".//xs:attribute[@name='coda']">
                <property name="codaAttribute"/>
            </bindings>
        </bindings>

        <bindings node="//xs:complexType[@name='credit']">
            <bindings node="./xs:sequence/xs:element[@name='link']">
                <property name="lnk"/>
            </bindings>
            <bindings node="./xs:sequence/xs:element[@name='bookmark']">
                <property name="bookmrk"/>
            </bindings>
        </bindings>
    </bindings>
</bindings>`

There is probably a better way to handle the "link" and "bookmark" items but this at least makes classes.

I also had to modify the musicxml.xsd file to use local xml.xsd and xlink.xsd because the Java/1.8 HTTP user agent was being blocked by [www.musicxml.org](musicxml.org) (Edit: turns out Cloudflare identifies the Java user agent used by XJC as a crawler or bot and blocks it) causing an issue where XJC could not retrieve the namespace files. Verified using `curl -A "Java/1.8" -I http://www.musicxml.org/xsd/xml.xsd` returns a 403 while `curl -A "Mozilla/5.0 (Windows NT 10.0; Win64; x64)" -I https://www.musicxml.org/xsd/xml.xsd` returns a 200. There might be a way to set the XJC user agent but since the local files are there I went that route instead.

**Comment by mdgood** [comment]
Thanks @dancarpenter21! This seems to address all the issues in #256, and appears to be modeled on the StackOverflow solution referenced in that issue.

Have you tried running xjc on the current MusicXML 4.0 schema? Are there any additional problems beyond these?

Also, would using the XML catalog provided in MusicXML 4.0 eliminate the need to modify the musicxml.xsd file? The catalog should be setup to resolve to your local version and not try to access the net version. Does xjc make use of XML catalogs if they are available?

**Comment by mdgood** [comment]
Does anybody else have solutions they are willing to share for other tools than generating JAXB binding with xjc? I think including even just the xjc approach would be helpful, but it would be even better if we could reference more tools. We are looking for solutions that work without making changes to the musicxml.xsd schema.

Also, where in the documentation do you think would be a good place for this information? My first thought is to add a new section to the tutorial, but other ideas are most welcome.

@dancarpenter21, could you please join the W3C Music Notation Community Group, unless you already are a member and I missed you in the list? You can join by going to https://www.w3.org/community/music-notation/ and clicking on the JOIN OR LEAVE THIS GROUP button. This will guide you through creating a W3C account and then joining the group. Membership is free of charge.

**Comment by dancarpenter21** [comment]
> Thanks @dancarpenter21! This seems to address all the issues in #256, and appears to be modeled on the StackOverflow solution referenced in that issue.
> 
> Have you tried running xjc on the current MusicXML 4.0 schema? Are there any additional problems beyond these?
> 
> Also, would using the XML catalog provided in MusicXML 4.0 eliminate the need to modify the musicxml.xsd file? The catalog should be setup to resolve to your local version and not try to access the net version. Does xjc make use of XML catalogs if they are available?

I'm not an expert on XML schemas and Java bindings, I'm not sure what "xml catalog" means. I assume it's the xml.xsd and link.xsd files:

The same approach to solving XJC issues in 3.1 were necessary and sufficient in 4.0. I think accessing the net distributions of xml.xsd and xlink.xsd would be fine if the Cloudflare proxy wasn't blocking my XJC tool from accessing them. I assume other tools for other languages would work fine; it doesn't seem that worthwhile to change the entire system simply to enable a single tool.

**Comment by mdgood** [comment]
Thanks again @dancarpenter21. XML catalogs are a way to handle this type of locally processing automatically. xjc offers a -catalog option, but the implementation is incomplete and doesn't work with MusicXML's catalog.xml file. So I have added this to the tutorial chapter. We added XML catalog support to version 4.0 in #259.

The chapter is now up for review at https://w3c.github.io/musicxml/tutorial/code-generation/. It focuses on problems with first-time generation of code from the MusicXML schema, rather than problems updating from older MusicXML versions to version 4.0.

Suggestions are most welcome! I used xjc for the detailed example and included the CodeSynthesis tools and the Python xsData tool as other code generator examples.
